File: leetcode/0200-number-of-islands/solution.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int islands = 0;

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                //found an island
                if(grid[i][j] == '1'){
                    islands++;
                    dfs(grid, i, j);
                }
            }
        }
        return islands;
    }

    private void dfs(char[][] grid, int i, int j){
        if(i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] == '0')
            return;
        
        //mark as visited by switching to 0
        grid[i][j] = '0';
        int[][] directions = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
        for(int[] direction : directions){
            int newI = i + direction[0];
            int newJ = j + direction[1];
            dfs(grid, newI, newJ);
        }
        return;
    }
}

File: leetcode/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        islands = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j, m, n)
                    islands += 1
        
        return islands
    
    def dfs(self, grid: List[List[str]], i: int, j: int, m: int, n: int) -> None:
        if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] == '0':
            return None
        grid[i][j] = '0'

        self.dfs(grid, i + 1, j, m, n) # check below
        self.dfs(grid, i - 1, j, m, n) # check above
        self.dfs(grid, i, j + 1, m, n) # check right
        self.dfs(grid, i, j - 1, m, n) # check left

        return



